numalogic/postprocess.py

```diff
@@ -9,24 +9,165 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-
 import numpy as np
-from numpy.typing import ArrayLike
+import numpy.typing as npt
 
 from numalogic.tools import DataIndependentTransformers
+from numalogic.tools.exceptions import InvalidDataShapeError
+
+
+def _allow_only_single_feature(data: npt.NDArray[float]) -> None:
+    if data.ndim > 2:
+        raise InvalidDataShapeError(
+            f"Input data can only be 2 dimensions or less, input size: {data.shape}"
+        )
+    if data.ndim > 1 and data.shape[1] > 1:
+        raise InvalidDataShapeError(
+            f"Input data can only have 1 feature column, input shape: {data.shape}"
+        )
 
 
-def tanh_norm(scores: ArrayLike, scale_factor=10, smooth_factor=10) -> ArrayLike:
+def tanh_norm(scores: npt.NDArray[float], scale_factor=10, smooth_factor=10) -> npt.NDArray[float]:
     return scale_factor * np.tanh(scores / smooth_factor)
 
 
+def expmov_avg_aggregator(
+    arr: npt.NDArray[float], beta: float, bias_correction: bool = True
+) -> float:
+    """
+    Aggregate a window of data into an expoentially weighted moving average value.
+
+    V(n) = (1 - beta) * beta**n * sum(x(i)/beta**i)   [for i = 1 to i = n]
+
+    "1.0 - beta" denotes the weight given to the latest element.
+
+    Args:
+        arr: single feature numpy array
+        beta: how much weight to give to the previous weighted average (n-1)th value
+        bias_correction: flag to perform bias correction (default: true)
+
+    Raises:
+        ValueError: if beta is not between 0 and 1
+        InvalidDataShapeError: if input array is not single featured
+    """
+    if beta < 0.0 or beta > 1.0:
+        raise ValueError("beta only accepts values between 0 and 1")
+    _allow_only_single_feature(arr)
+
+    # alpha is the weight given to the latest element
+    alpha = 1.0 - beta
+    n = len(arr)
+    theta = arr.reshape(1, -1)
+    powers = np.arange(1, n + 1)
+
+    # Calculate increasing powers of beta of the form [beta, beta**2, .., beta**n]
+    beta_powers = np.power(beta, powers)
+
+    # Calculate the reciprocals of beta powers
+    # [beta**(-1), beta**(-2), .., beta**(-n)]
+    beta_arr_inv = np.reciprocal(beta_powers.reshape(-1, 1))
```

Comment:
it'd be simpler to just do the following, and skip the reciprocal step.
```
powers = np.arange(n-1, 0, -1)
beta_powers = np.power(beta, powers)
```

Reply:
Acknowledged. Fixing it.